module-4/backend/week_two.md

```diff
@@ -1,22 +1,38 @@
+  Synchronous JavaScript functions in an absolutely single-threaded manner, meaning that each code process accumulates on the stack in a linear order and must be resolved sequentially, even if a function makes requests that are blocking and require the program to wait. In the browser, JavaScript uses a browser engine to replicate multi-threaded behavior by placing functions that make an external request to an API onto the stack through an additional channel. This allows JavaScript to execute other things while fetch requests etc are made.
 3. What are the four pillars of Object Oriented programming?
+  Abstraction: We abstract methods into small pieces of functionality that can be easily tested to verify their functionality. We then use these abstracted pieces within other functions to abstract way the complexity of various pieces of our code.
+  Encapsulation: encapsulation refers to the coupling of functionality and data storage within single objects. We store data within methods and also dictate behavior for that data.
+  Polymorphism: Creating methods that appear to do the same thing to different pieces of data, but actually do very different things on the good. For example, .length on strings and .length on arrays.
+  Inheritance: We can create hierarchies of inheritance that allow us to build only the exact functionality that we want in each object while not repeating ourselves.
 4. What are some tools available in JavaScript to help you write object oriented code?
+  We can use constructors and prototypes to write this code. Constructors allow us to create objects with various properties, and prototypes allow us to pass along functionality without the data bloat of storing behavior on individual objects.
 5. What are some key concepts to be aware of when refactoring your JavaScript?
+  Method length, duplicate code, complex conditional logic, long parameters, obscuring comments.
 6. What's a callback function and what are some reasons when we use/need callback functions?
+  Callback functions are functions that are passed into other functions as arguments. This is possible because functions are first-class objects in JavaScript.
 7. What are the different scopes of variables in Javascript? When would you want to define global variables?
+  var is a variable that is available in the scope it is defined. Originally, javascript did not support block scoping, so a var defined in a block would be available outside of that block. Let is a variables that is block scoped. Const is a var that cannot be changed.
 8. What's the difference between `==` and `===` in JavaScript?
+  they are loose and strict equality, respectively. JavaScript will attempt to coerce datatypes into something common to compare things, such as if we were to compare 2 and "2". Strict equality does not allow for that coercion.
 9. Why do front end frameworks exist?
-
-#### Review  
-
+  These exist to provide us with ways to organize our files.
```

Comment:
They also DO a lot for us - they bundle up the basic functionality that devs need, in an easier-to-work-with way! Libraries like React do that as well, so the 'helpful' aspect isn't exclusive to frameworks.